src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/markduplicates/MarkDuplicatesUtils.java

```diff
@@ -216,6 +266,209 @@ public void processElement(final ProcessContext context) throws Exception {
                 });
     }
 
+    // This transform takes marked reads and generates the metrics from those tests.
+    public static class GenerateMetricsTransform extends PTransform<PCollection<GATKRead>, PCollection<KV<String, DuplicationMetrics>>> {
+        private static final long serialVersionUID = 1l;
+
+        /**
+         *  The header for all reads processed.
+        */
+        private final PCollectionView<SAMFileHeader> header;
+
+        public GenerateMetricsTransform(final PCollectionView<SAMFileHeader> header) {
+            this.header = header;
+        }
+
+        @Override
+        public PCollection<KV<String, DuplicationMetrics>> apply(final PCollection<GATKRead> preads) {
+            return preads
+                .apply(generateMetricsByLibrary())
+                .setCoder(KvCoder.of(StringUtf8Coder.of(), new DuplicationMetricsCoder()))
+                // Combine metrics for reads from the same library.
+                .apply(Combine.<String, DuplicationMetrics>perKey(new CombineMetricsFn()))
```

Comment:
Dataflow claims to be smart enough to handle this:
https://cloud.google.com/dataflow/model/combine
"When you design a combine function, note that the function is not necessarily invoked exactly once on all values with a given key. Because the input data (including the value collection) may be distributed across multiple worker instances, the combining function might be called multiple times to perform partial combining on subsets of the value collection. Generally speaking, the combine might be repeatedly applied in a tree structure. Because the tree structure is unspecified, your combining function should be associative."